power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,377 @@
+
+class Deserializer {
+  public:
+    // struct of buffer data
+    struct Buffer {
+        MetaComponent const* component;
+        // for non-uniform component, this is -1, we use indptr to describe the elements per scenario
+        Idx elements_per_scenario;
+        Idx total_elements;
+        std::vector<std::span<msgpack::object const>> msg_data; // vector of spans of msgpack object of each batch
+        void* data;                                             // set by user
+        std::span<Idx> indptr;                                  // set by user
+    };
+
+    Deserializer() = default;
+    // not copyable
+    Deserializer(Deserializer const&) = delete;
+    Deserializer& operator=(Deserializer const&) = delete;
+
+    void deserialize_from_json(char const* json_string) {
+        std::vector<char> const msgpack_data = json_to_msgpack(json_string);
+        deserialize_from_msgpack(msgpack_data.data(), msgpack_data.size());
+    }
+
+    void deserialize_from_msgpack(char const* data, size_t length) {
+        handle_ = msgpack::unpack(data, length);
+        post_serialization();
+    }
+
+    std::string const& dataset_name() const { return dataset_->name; }
+
+    bool is_batch() const { return is_batch_; }
+
+    Idx batch_size() const { return batch_size_; }
+
+    Idx n_components() const { return (Idx)buffers_.size(); }
+
+    Buffer const& get_buffer_info(Idx i) const { return buffers_[i]; }
+
+    void set_buffer(char const** components, void** data, Idx** indptrs) {
+        for (Idx i = 0; i != n_components(); ++i) {
+            auto const found = std::find_if(buffers_.begin(), buffers_.end(), [components, i](Buffer const& buffer) {
+                return buffer.component->name == components[i];
+            });
+            if (found == buffers_.end()) {
+                throw SerializationError{"Unkown component: " + std::string(components[i]) +
+                                         "! You need to supply the components which are present.\n"};
+            }
+            found->data = data[i];
+            if (found->elements_per_scenario < 0) {
+                found->indptr = {indptrs[i], (size_t)(batch_size_ + 1)};
+            }
+        }
+    }
+
+    void parse() const {
+        for (Buffer const& buffer : buffers_) {
+            parse_component(buffer);
+        }
+    }
+
+  private:
+    msgpack::object_handle handle_{};
+    std::string version_;
+    bool is_batch_{};
+    MetaDataset const* dataset_{};
+    std::map<std::string, std::vector<MetaAttribute const*>> attributes_;
+    Idx batch_size_{}; // for single dataset, the batch size is one
+    std::vector<Buffer> buffers_;
+    // attributes to track the movement of the position
+    // for error report purpose
+    mutable std::string_view root_key_;
+    mutable std::string_view component_key_;
+    mutable std::string_view attribute_key_;
+    mutable Idx scenario_number_{-1};
+    mutable Idx element_number_{-1};
```

Comment:
i'm not really a fan of `mutable` keywords. they may result in unexpected behaviour when dealing with `const` functions. I'd rather provide some kind of error metadata target stream and log error messages there than doing it the other way around

Reply:
These member are the error metadata target stream.

The error metadata target stream has to live somewhere, either as mutable class members, or passing around by all the functions. I prefer the first one. We can even remove the `mutable` keywords and make all the functions non-const. For deserializer class we don't actually need const function.

Reply:
we can maybe make the best of both worlds by adding a private `struct Cache` that we can pass around.

then we only need to pass around a single object with multiple attributes instead of a separate object for every attribute that bloats the argument list. That way we can keep the functions `const` but still make very clear to the users that there's a cache that captures the latest state

otherwise, i'm in favour of removing the `const` specifier.

Reply:
`const` functions removed, and `mutable` removed.